recipe/2024-04-terraform-provider-secrets.md

```diff
@@ -0,0 +1,260 @@
+              java_endpoint: '...'
+              mysql_endpoint: '...'
+            }
+          ]
+        }
+      }
+      env: {
+        MY_ENV_VAR_1: 'my_value'
+      }
++      envSecrets: { // Individual Secrets from SecretStore
++        MY_ENV_VAR_2: {
++          source: secretStoreConfig.id
++          key: 'envsecret.one'
++        }
++        MY_ENV_VAR_3: {
++          source: secretStoreConfig.id
++          key: 'envsecret.two'
++        }
++     }
+    }
+    recipes: {
+    }
+  }
+}
+
+```
+
+**Sample Output:**
+n/a
+
+**Sample Recipe Contract:**
+n/a
+
+## Design
+
+### High Level Design
+
+The design for handling secrets for Private Terraform modules established the basis for data flow between Engine and Driver components in the codebase. For Terraform Providers, we will continue to maintain the flow where secrets are retrieved in the Engine and passed into the Driver in the Execute() call. 
+The following diagram describes the updated data flow. The calls in green highlight updates to existing code.
+
+![alt text](2024-04-terraform-provider-secrets.png)
+
+### Detailed Design
+
+The FindSecretIds function in the DriverWithSecrets interface is updated to return a map of SecretStoreIds and keys. These will hold data for a Terraform module as before and in addition, also hold SecretStoreIds and keys for provider configuration and environment variables, iterating through the input environment recipe configuration. 
+
+The LoadSecrets method in the SecretLoader interface will be updated return secret data for multiple Ids returned from FindSecretIds().
+
+#### Advantages (of each option considered)
+Retrieving secrets in the Engine and passing them into the Driver was discussed positively as a way forward. Looking ahead with Containerization, the Driver will have all the information it needs to be able to be spin off into a separate process.
+
+#### Disadvantages (of each option considered)
+N/A
+
+#### Proposed Option
+N/A
+
+### API design (if applicable)
+N/A
+
+### CLI Design (if applicable)
+N/A
+
+### Implementation Details
+###Recipes (Engine and Driver)
+
+
+Driver Updates:
+```go
+//*** CURRENT ***//
+
+type DriverWithSecrets interface {
+  ...
+	// FindSecretIDs gets the secret store resource ID references associated with git private terraform repository source.
+	// In the future it will be extended to get secret references for provider secrets.
+	FindSecretIDs(ctx context.Context, config recipes.Configuration, definition recipes.EnvironmentDefinition) (string, error)
+}
+
+//*** NEW ***//
+
+// FindSecretIDs will be updated in current DriverWithSecrets interface to now return map of secretStoreIds and keys
+type DriverWithSecrets interface {
+  ...
+	// FindSecretIDs gets the secretStore resource IDs and keys for the recipe including module, provider, envSecrets.
+	FindSecretIDs(ctx context.Context, config recipes.Configuration, definition recipes.EnvironmentDefinition) (map[string]string, error)
```

Comment:
What is the key of of the map and what is the value? I know it's strings, but what do they represent?

Reply:
Corrected it: updated to map[string][]string // map[SecretStoreId]ListOfKeysInSecretStoreId.